src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -3812,6 +3812,20 @@ class CommonVisitor : public AST::BaseVisitor<Struct> {
                 tmp = make_TupleItem_t(al, loc, value, index,
                                        ASR::down_cast<ASR::Tuple_t>(type)->m_type[i], nullptr);
                 return false;
+            } else if (ASR::is_a<ASR::Array_t>(*type)) {
+                index = ASRUtils::EXPR(tmp);
+                ASR::expr_t* val = ASRUtils::expr_value(index);
+                if (val && ASR::is_a<ASR::IntegerConstant_t>(*val) && 
+                    ASRUtils::extract_physical_type(type) != ASR::array_physical_typeType::DescriptorArray) {
+                    if (ASR::down_cast<ASR::IntegerConstant_t>(val)->m_n < 0) {
+                        ASR::ttype_t *int_type = ASRUtils::TYPE(ASR::make_Integer_t(
+                                                        al, loc, 4));
+                        ASR::expr_t *neg_idx = ASRUtils::expr_value(index);
+                        ASR::expr_t *dim_size = ASR::down_cast<ASR::Array_t>(type)->m_dims[idx].m_length;
```

Comment:
This gives you the dimension size only when it is available at compile-time. For descriptor arrays, the length is available at runtime. For example, the following fails with the changes in this PR.

```console
% cat examples/expr2.py   
from lpython import i32
from numpy import empty, int32, array

def foo(x: i32[:]):
    print(x[3], x[4], x[-1], x[-2])

def main():
    x: i32[5] = empty(5, dtype=int32)
    x = array([1, 2, 3, 4, 5])
    foo(x)

main()
% python examples/expr2.py
4 5 5 4
% lpython examples/expr2.py
4 5 1 12844716
```

To get the dimension size at runtime, you can use `ArrayBound()` and pass it the appropriate dimension index (whose dimension you need).

Reply:
```console
% cat examples/expr2.py 
from lpython import i32
from numpy import empty, int32, array

def foo(x: i32[:]):
    print(x[3], x[4], x[-1], x[-2])

def main():
    x: i32[5] = empty(5, dtype=int32)
    x = array([1, 2, 3, 4, 5])
    foo(x)

main()
% python examples/expr2.py 
4 5 5 4
% lpython examples/expr2.py
4 5 5 4
```
The above test case now works.